### PR TITLE
Moez/ruberics read me

### PR DESCRIPTION
**Pull request by moezthedev**
# Pull Request Template

Before submitting this PR, please make sure you have completed the following checklist. This helps ensure that our codebase remains high quality and that our features work well across all supported devices and themes.

### Code Review
- [x] I have carefully reviewed my code and believe it is final from my side.
- [x] I have ensured there is no hardcoded color in my changes. Colors should be defined in theme files and reused.

### Theme Compatibility
- [x] I have tested the new feature or updates in both light and dark themes.
- [x] I confirm that the visual aspects of the feature are compatible and appealing in both themes.

### Device Compatibility
- [x] I have tested my changes on mobile screen sizes.
- [x] I have tested my changes on tablet screen sizes.
- [x] I have tested my changes on laptop screen sizes.

### Functional Testing
- [x] I have thoroughly tested the functionality of my changes.
- [x] I am confident that my changes do not negatively impact existing features.

### Testing Documentation
Please provide a brief description of the tests you have performed to verify your changes. Include any relevant scenarios or edge cases.

- Example test 1: Tested feature X by uploading the image
- Example test 2: Tested feature X by doing these things.

# Demo Video
Please provide a demo video of your changes. This helps reviewers understand the changes and verify the functionality.

You can record the video using tools like Loom(https://www.loom.com/), Clip by ClickUp(https://clickup.com/features/clips), Sendspark(https://www.sendspark.com/), etc.

Include the link to the video below and make sure the video covers
- [x] Explain the task/bug you are solving
- [x] Show the changes in the UI
- [x] Show the changes in the codebase
- [x] Show the changes you tested to make sure it works
- [x] Show the changes in various screen sizes
- [x] Show the changes in both light and dark themes
Videos: 
1) https://www.loom.com/share/bb0963dd4fda4478ba5bbb9b0fcd2018
2) https://www.loom.com/share/56969bd25589453cb9203d7d3eb8bfe0
3 )https://www.loom.com/share/96c9155e0c84467d9dd6f0152fd27716

### CODE REVIEW

File: rubrics-ui/src/app/api/ruberics/create-programs/route.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { NextApiResponse } from 'next';
+import { PrismaClient } from '@prisma/client';
+import { NextRequest, NextResponse } from 'next/server';
+const prisma = new PrismaClient();
+
+export async function POST(req: NextRequest, res: NextApiResponse) {
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-programs/route.ts:6]: make the type of res consistent with NextResponse

File: rubrics-ui/src/app/api/ruberics/create-programs/route.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  if (req.method === 'POST') {
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-programs/route.ts:7]: We don't need to check the request type

File: rubrics-ui/src/components/rubricsTable/rubricsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { RubricsPageProps } from '../../types/rubricsTypes/types';
```
Comment (MSamiTariq) [inline rubrics-ui/src/components/rubricsTable/rubricsTable.tsx:1]: We don't usually specify the path this way, see examples in other files

File: rubrics-ui/src/app/api/ruberics/create-programs/route.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+      const newProgram = await prisma.program.create({
+        data: {
+          name,
+          details,
+          rating,
+        },
+      });
+
+      return NextResponse.json({ status: '200', body: newProgram });
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-programs/route.ts:23]: status should be a number, not a string

File: rubrics-ui/src/app/api/ruberics/create-programs/route.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+const prisma = new PrismaClient();
+
+export async function POST(req: NextRequest, res: NextApiResponse) {
+  if (req.method === 'POST') {
+    try {
+      const { name, details, rating }: { name: string; details?: string; rating: number } = await req.json();
+
+      if (!name) {
+        return new Response('Missing required fields');
+      }
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-programs/route.ts:13]: Only name field is mandatory? I think we should handle this case on the front end. If any field is missing, return a 404 error

File: rubrics-ui/src/app/api/ruberics/create-programs/route.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        },
+      });
+
+      return NextResponse.json({ status: '200', body: newProgram });
+    } catch (error) {
+      console.error('Error creating program:', error);
+      return NextResponse.json({ status: '500', body: 'Failed to create a response' });
+    }
+  } else {
+    return NextResponse.json({ status: '405', body: 'Method not allowed' });
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-programs/route.ts:29]: I don't think we need this either

File: rubrics-ui/src/app/api/ruberics/create-programs/route.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import { NextApiResponse } from 'next';
+import { PrismaClient } from '@prisma/client';
+import { NextRequest, NextResponse } from 'next/server';
+const prisma = new PrismaClient();
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-programs/route.ts:4]: initialize the prisma client in a separate file and then import it, else this will error out with too many prisma clients. See academy-ui for inspiration on how we do this

File: rubrics-ui/src/app/api/ruberics/create-rubrics/route.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import { PrismaClient } from '@prisma/client';
+import { NextRequest, NextResponse } from 'next/server';
+
+const prisma = new PrismaClient();
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-rubrics/route.ts:4]: Same goes here, initialize in a separate file

File: rubrics-ui/src/app/api/ruberics/create-rubrics/route.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  try {
+    const { name, summary, description, criteria, programId, evaluationParameters } = await req.json();
+
+    if (!name || !summary || !criteria || !programId || !evaluationParameters) {
+      return NextResponse.json({ error: 'Missing required fields' }, { status: 400 });
+    }
+
+    if (!Array.isArray(evaluationParameters) || evaluationParameters.length === 0) {
+      return NextResponse.json({ error: 'Evaluation parameters are required and must be an array' }, { status: 400 });
+    }
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-rubrics/route.ts:16]: This should be handled on the front end

File: rubrics-ui/src/app/api/ruberics/create-rubrics/route.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    for (const param of evaluationParameters) {
+      if (!param.title || !param.description) {
+        return NextResponse.json({ error: 'Each evaluation parameter must have a title and description' }, { status: 400 });
+      }
+    }
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/api/ruberics/create-rubrics/route.ts:22]: same goes for this, handle this on front end

File: rubrics-ui/src/app/editrubrics/page.tsx [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import ProgramList from '../../components/programList/programList';
```
Comment (MSamiTariq) [inline rubrics-ui/src/app/editrubrics/page.tsx:2]: correct the path

File: rubrics-ui/src/app/editrubrics/page.tsx
Comment (MSamiTariq) [inline rubrics-ui/src/app/editrubrics/page.tsx:1]: folder name should be 'edit-rubrics'

**Review by MSamiTariq — CHANGES_REQUESTED** [review]

File: rubrics-ui/src/components/programList/programList.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  const handleCreateProgramClick = async () => {
+    if (!newProgram.name || !newProgram.details) {
+      setError('Please fill in all fields with valid values.');
+      return;
+    }
+
+    setError(null);
+
+    try {
+      const response = await fetch('/api/ruberics/create-programs', {
```
Comment (MSamiTariq) [inline rubrics-ui/src/components/programList/programList.tsx:30]: endpoint should be called `/create-program` as it creates a single program only

File: rubrics-ui/src/components/programList/programList.tsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newProgram),
+      });
+
+      setNewProgram({ name: '', details: '' });
+      router.push('/edit-rubrics');
+      router.refresh();
```
Comment (MSamiTariq) [inline rubrics-ui/src/components/programList/programList.tsx:40]: we should look for some way to fetch the programs instead of refreshing the page

File: rubrics-ui/src/types/rubricsTypes/types.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  id: number;
+  name: string;
+  details?: string | null;
+  rating: number;
+  rubrics: Rubric[];
+}
+export interface ServerResponse {
+  status: number;
+  body: Program[];
+}
```
Comment (MSamiTariq) [inline rubrics-ui/src/types/rubricsTypes/types.ts:11]: maybe rename this interface to something specific to Program

File: rubrics-ui/src/components/rubricsTable/rubricsTable.tsx [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+      case 0:
+        return 'bg-green-500';
+      case 1:
+        return 'bg-yellow-400';
+      case 2:
+        return 'bg-yellow-600';
+      case 3:
+        return 'bg-red-500';
+      default:
+        return 'bg-gray-400';
```
Comment (MSamiTariq) [inline rubrics-ui/src/components/rubricsTable/rubricsTable.tsx:178]: confirm with Robin if these should be hard coded

File: rubrics-ui/src/components/rubricsTable/rubricsTable.tsx [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+            </tr>
+          </thead>
+          <tbody>
+            {criteriaOrder.map((criteria) => (
+              <tr key={criteria}>
+                <td className="py-2 px-4 border-r border-b font-bold cursor-pointer max-w-xs break-words relative">
+                  <div onClick={() => handleEditClick('criteria', criteria, -1)} className="overflow-y-auto max-h-24">
+                    {criteria}
+                  </div>
+                  <button onClick={() => handleDeleteCriteria(criteria)} className="absolute top-0 right-0 bg-red-500 text-white p-0.5 rounded-half">
```
Comment (MSamiTariq) [inline rubrics-ui/src/components/rubricsTable/rubricsTable.tsx:218]: you can use a trash bin icon for delete, and the delete icon position to may be at the end of the row

**Review by MSamiTariq — CHANGES_REQUESTED** [review]

File: rubrics-ui/prisma/migrations/20240715174819_initial/migration.sql [hunk 17]
```diff
@@ -74,6 +75,58 @@ CREATE TABLE "spaces" (
...
+    CONSTRAINT "Program_pkey" PRIMARY KEY ("id")
+);
+
+-- CreateTable
+CREATE TABLE "Rubric" (
+    "id" TEXT NOT NULL,
+    "name" TEXT,
+    "summary" TEXT,
+    "description" TEXT,
+    "programId" TEXT NOT NULL,
```
Comment (RobinNagpal) [inline rubrics-ui/prisma/migrations/20240715174819_initial/migration.sql:94]: We will not have programId in Rubric. You can remove this column for now

Comment (RobinNagpal) [reply rubrics-ui/prisma/migrations/20240715174819_initial/migration.sql:94]: We will have a mapping table which we can add in the next PR
